### ISSUE TITLE
Integrate e-mission and itinerum (finally!)

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#643]
@PatGendre @asiripanich you should be happy about this :)

@kafitz and I just had the preliminary discussion around how the integration might work.
A very basic data-level (not code-level) integration would be:
- e-mission, on startup, pulls the survey JSON from the itinerum server
   - the survey JSON access code does not have authentication, so this is easy
- e-mission creates a locally generated token (similar to NREL-Lo-Hi and @asiripanich's code) which will be used to log in to the e-mission server
- itinerum will create their own instances of the e-mission server and database containers running on their own server
- we will write two integration containers:
   - one which reads the e-mission data and exposes it via a simple API
   - one which formats the data into Postgres and saves it into the Itinerum database
- the exporter container will not be externally exposed, so the API does not need authentication
- there will be a separate "EXPORT_DATA" pipeline stage which will be maintained similar to the other pipeline stages, and so will support incremental updates

That seems eminently doable in the short internship timeframe

**Comment by shankari** [comment]
@kafitz one thing that we are still missing is displaying the survey after we retrieve it. Do you have a javascript library to do that (maybe for a preview)? could you add the link here as well?

**Comment by kafitz** [comment]
This is something I meant to ask about this on our call today, so far that's been handled by the native mobile Itinerum apps. Currently we can see and edit the questions using the existing dashboard, but the client view for e-mission would still need to be built.

Is the best way to write some HTML/JS viewer now that could eventually be merged into the e-mission app? It might take a week or so (for a prototype), but seems doable relatively quickly.

**Comment by shankari** [comment]
@kafitz yes, a HTML+JS viewer that would be merged with the e-mission app would be the way to go. I was assuming you would have such a viewer anyway as part of the survey builder to preview the survey as it was being built. How do survey users preview the survey right now?

**Comment by kafitz** [comment]
Currently we just rely on the Survey Builder wizard for laying out questions and then the admin user has tested out how it looks by running the app. If any survey questions must be fixed, it does mean they have to clear the app data and retry.

The builder provides a decent enough overview for this, but it's written in React. Since the app is Cordova/Angular (which I'm still not super familiar with), I was thinking now I can create a very simple interface with HTML & vanilla JS and this will hopefully keep it easy to merge into the e-mission-phone code later.

Open to suggestions ofc if there's a better way to do it. I'll admit I'm slightly worried about it snowballing, but I think it should be ok if we don't worry about aesthetics until putting it into the app.

**Comment by shankari** [comment]
so I've been poking around with the edges of react for a while (https://github.com/NREL/mobility_landscapeapp is in React, for example), and it looks like you can use some form of react embedded into other applications.
https://reactjs.org/docs/getting-started.html
or
https://stackoverflow.com/questions/20947191/using-reactjs-with-angularjs

I can play around with it a bit more; if it works it seems like it would be the easiest option without introducing additional codebases that need to be maintained. It would also give us a more gradual migration path forward - as we cleanup individual components, we can rewrite them in react.

Is there a simple enough react component that you can share which I can use to experiment with this approach?

**Comment by shankari** [comment]
This looks super simple, for example
https://mtalham.medium.com/build-web-component-with-react-and-use-it-in-an-angular-app-515c2fed6a53

Famous last words 😄

**Comment by kafitz** [comment]
The Tutorial.jsx snippet on this gist page is my most basic component for the help screen users first see in the dashboard. 
https://gist.github.com/kafitz/3f4995ce85e59a346f94db2bbf37bc27

I can send a more involved example, but that may take a bit since I rewrote the codebase using Typescript/Material UI. That code resembles something closer to React Native and requires transpiling...
<img width="590" alt="Screen Shot 2021-06-02 at 8 22 24 PM" src="https://user-images.githubusercontent.com/253651/120567663-46a15180-c3e0-11eb-887d-36b6500f1883.png">

For anything that would go in e-mission, I would still write that in regular React for simplicity.

**Comment by shankari** [comment]
I think we can get webpack transpiling if we need to as well.
(https://mtalham.medium.com/build-web-component-with-react-and-use-it-in-an-angular-app-515c2fed6a53) seems to have a standard React *app*

But let's start with seeing if I can embed this first.

**Comment by shankari** [comment]
@kafitz can you add a link to the API that we can use to pull the survey JSON (I assume this is on a staging server, and there is one canonical server we can pull). we can then start with reading that. At tomorrow's meeting, let's discuss at what stage on the onboarding process we should make this call and how the user specifies the survey to pull.

**Comment by shankari** [comment]
to give some additional context: e-mission is currently written in angular1 also called angularJS on the internet. It is also a reactive framework in which you make changes to `$scope` variables in a `controller` which are automatically updated in the HTML.

So concretely, if I want to display a username, I can have `<h3>{{username}}</h3>` in the HTML, set `$scope.username = "Jenna";` in the javascript controller and it will show up. It also has some additional tags to make the UI coding easier, so for example,  (syntax may be a bit off, high level workflow only)

```
<ion-list ng-repeat="trip in tripList">
    <ion-item>{{trip.start_fmt_time}} -> {{trip.end_fmt_time}}</ion-item>
</ion-list>
```

then if the javascript has `$scope.tripList = [{start_fmt_time: "...", end_fmt_time: "..."}, {...}, {...}, ...]` then the list will show all the trips.

**Comment by shankari** [comment]
once we get the survey JSON from itinerum, we want to display it. @kafitz already has code to display it that is written in React, which is a very similar-looking but different javascript framework. So we are going to try to package the react code as a component and include it in the angular framework. The comments above claim it can be done. If not, we may have to rewrite parsing + display logic in angular.

**Comment by kafitz** [comment]
The base url for the staging server will be at: https://api.hungry.wales/mobile/v2

Attached here is a shortened version of an integration testing script which includes two functions: one to make the initial API call for the survey JSON and one to POST the answers.

I recently started adding some Swagger docs which may give some idea of the requests, but there's a lot more work to do on that to make it clear: https://api.hungry.wales/mobile/swagger/index.html

Something to note is this would be the app to use my "v2" API which was built with conditional question logic in mind. It's backwards compatible with v1, but you'll the survey JSON itself is pretty meaty. I followed [JSON Schema](https://json-schema.org/) for declaring my survey model--personally I find their documentation difficult, but this tool might help show it better: https://react-jsonschema-form.readthedocs.io/en/latest/usage/dependencies/

[emission-integration-demo.zip](https://github.com/e-mission/e-mission-docs/files/6619461/emission-integration-demo.zip)
(sorry about the zip, Github issues doesn't like .py files I guess)

**Comment by kafitz** [comment]
Also I would say that the client-side React code is about 50% there. 

We have a question builder interface which parses the JSON and maps each question to components, but it will need transformed into a client version. This I was planning to help write for the UI integration, but haven't yet ruled out Angular here if that winds up making more sense.

![survey builder](https://user-images.githubusercontent.com/253651/121258123-85ba2180-c87c-11eb-8b67-bb85700b1900.PNG)

**Comment by kafitz** [comment]
To clarify about the requests for retrieving the survey, there will be two POST requests:

1. There will the initial POST request to send parameters about the phone, a UUID (generated on phone), and a survey name which will return the survey parameters and questions as JSON.

2. You then create a POST request with survey answers and the UUID.

**Comment by shankari** [comment]
Just to clarify how this would work on the e-mission side, because of our legacy of authenticating via email, we actually generate a random token first and "register" with the server, similar to step 1 in your case.
We then retrieve the UUID from the server and use it to fill out any surveys that are launched during onboarding. The surveys are currently not saved to e-mission since we don't have a survey builder, but we can fairly easily change that. We should then be able to send a POST request with the uuid and the survey result to the itinerum server.

@jruzekowicz the code to currently launch the survey with a UUID is at 
https://github.com/e-mission/e-mission-phone/blob/master/www/js/survey/launch.js

It is invoked from https://github.com/e-mission/e-mission-phone/blob/ceo_ebike_project/www/js/intro.js#L112 for a google form and from https://github.com/e-mission/e-mission-phone/blob/nrel_location_history/www/js/intro.js#L144 for a kobo toolbox form

I would suggest that an initial task would be:
- use one of those branches
- in SurveyLaunch, retrieve the sample JSON from @kafitz's example
- print it out on the console so we know it was retrieved
- send back a POST to @kafitz's server with the UUID and a example hardcoded filled out value

**Comment by shankari** [comment]
e-mission allows you to launch a survey on an end of trip prompt, but the general feedback we have gotten is that end of trip prompts are really intrusive.

So if deployments do want to have an end of trip prompt, right now, the default is to go to the diary screen where you can look at the trip and answer any "labeling" questions. That label can launch a survey, but people can wait until the end of the day to do it.

Most deployments have actually switched to once a day prompts, in the evening, that just launch the diary.

**Comment by shankari** [comment]
you would replace existing code with something like (very sketchy).

```
$scope.surveyLaunch = function(uuid) {
     const surveyJSON = $http.post(.....);
     console.log(surveyJSON);
     const surveyResult = ... hardcoded, copied from python printout....
     $http.post(.........surveyResult, uuid.....);
}
```

**Comment by shankari** [comment]
Looking through the various options above, there is a clear tradeoff between immediate usability and long-term migration. Concretely:
- the medium article: works with angular 11. Can we use a similar technique with angular1? Unclear
- `ngReact`: archived, not updated for two years. But the examples are very clearly angular1, the code is installed using bower, the imports are included as `<script` tags, etc. It suggest using `react2angular` or `ngImport`.
- `react2angular`: not archived, also updated 2 years ago. seems pretty similar to `ngReact`, but the examples have imports that look like this `import { react2angular } from 'react2angular'`. I am not sure how this will work with our current angular1 codebase. I do see this SO post (https://stackoverflow.com/questions/43137592/import-node-module-in-angularjs) which seems to indicate that these imports work for angular2, but the README says "Use it in your Angular 1 code"
- ngimport: recommended by `ngReact`, this really seems to be focused around "ugly, proprietary DI". It also provides "an easy migration path to angular2, React, etc". Note that DI stands for dependency injection - this is what it looks like in angular1. This seems orthogonal to our problem of including React components although it will help with longer-term migration.

**Comment by shankari** [comment]
at this point, given the tight timeframe and importance of getting something to work, I am going to start with ngReact. Depending on how long that takes to get working, I will try out `react2angular` or the medium article, otherwise, we are going with `ngReact`.

**Comment by shankari** [comment]
Starting with ngReact....
- the examples all seem to work with `bower`, so using it instead of `npm` for the install
- after cloning directly, and using `npx bower install ngReact`, I can run the first static hello world!!
   - Note that the angular component has `$scope.person` defined, and it passes it in to the react component as `props`
   - the react component uses `prop` throughout
   - quick check: is the binding bi-directional? let's add an angular-based HTML using `$scope` to figure out. Works!

```
diff --git a/examples/hello/index.html b/examples/hello/index.html
index 1b41b94..6c39b6b 100644
--- a/examples/hello/index.html
+++ b/examples/hello/index.html
@@ -31,6 +31,12 @@
   </head>
   <body ng-app="app" ng-controller="mainCtrl">
     <div class="content">
+        <h1>
+          <div>Hello {{person.fname}} {{person.lname}}</div>
+          <small>
+            <code>&lt;From angular to check bi-directional binding/&gt;</code>
+          </small>
+        </h1>
       <div class="header">
         <h1>
           <div><react-component name="Hello" props="person"/></div>
```

<img width="810" alt="Screen Shot 2021-07-11 at 10 52 32 AM" src="https://user-images.githubusercontent.com/2423263/125205236-1ac28880-e236-11eb-970e-de3283ca6aa9.png">

**Comment by shankari** [comment]
So it works for some version of react. Now, let's check the jsx stuff since that is what Kyle's example uses. That looks like it basically just includes `    <script src="../../bower_components/react/JSXTransformer.js"></script>`  and then just includes the jsx file directly. Let's try https://github.com/ngReact/ngReact/blob/master/examples/jsx-transformer/index.html as a static file.

**Comment by shankari** [comment]
That does not appear to work right out of the box. The failure is

```
ReferenceError: Hello is not defined
    <anonymous> file:///Users/kshankar/OSS/ngReact/examples/jsx-transformer/app.js:9
    Angular 27
    <anonymous> file:///Users/kshankar/OSS/ngReact/examples/jsx-transformer/app.js:12
angular.js:15697:41
```

Trying to understand `JSXTransformer.js` better...

**Comment by shankari** [comment]
Aha! JSXTransformer was deprecated in `0.13.x` and removed in `0.14.x`
https://reactjs.org/blog/2015/06/12/deprecating-jstransform-and-react-tools.html

the current version of ngReact uses react 15+, so it looks like the example was not updated. we can still use react and jsx without using webpack by including babel directly. babel can manually convert all jsx to js ahead of time.
https://reactjs.org/docs/add-react-to-a-website.html#optional-try-react-with-jsx

That approach works (see screenshot below)! 
<img width="951" alt="Screen Shot 2021-07-11 at 11 26 15 AM" src="https://user-images.githubusercontent.com/2423263/125206206-d2599980-e23a-11eb-99c4-da988649ae67.png">

We can now include jsx for sure, even if it is not super production ready. A minor caveat was that since we split the file into multiple javascript/jsx files, it ran into CORS issues. Note that loading from files doesn't work with CORS any more,
> [Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at .... (Reason: CORS request not http)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSRequestNotHttp)

so we have to start up a tiny web server. I used https://github.com/danjellesma/http-server; the same as em-dashboard and everything worked great.

**Comment by shankari** [comment]
Now, let's try the "production" version of babel without going all the way to webpack and transpiling. Although perhaps we should/will get there eventually.

**Comment by shankari** [comment]
huh! After copying it over, the "production" version works just fine. I guess I wasn't commenting something out properly before?

I can include the converted js file directly, and remove the manual bootstrap hack in `app.js` and everything Just Works

<img width="1323" alt="Screen Shot 2021-07-11 at 5 05 13 PM" src="https://user-images.githubusercontent.com/2423263/125214120-2b8cf180-e26a-11eb-953e-7c957bdffd30.png">

**Comment by shankari** [comment]
I'm now going to try to use this itinerum example with ngReact. Then last step is to use it with e-mission-phone.
Will deal with webpack etc after discussing packaging and repo structure with @kafitz

**Comment by shankari** [comment]
It (sort of) works.

<img width="1319" alt="Screen Shot 2021-07-11 at 5 38 44 PM" src="https://user-images.githubusercontent.com/2423263/125215019-dd2e2180-e26e-11eb-8afb-4a92888b9e99.png">

 I had to remove all the parts with @import and export because otherwise, even with the autogenerated js file, I got errors like 

```
Uncaught SyntaxError: export declarations may only appear at top level of a module
```

because the generated js file had 

```
export default TutorialComponent;
```

at the end.

Similarly, for the import statements, we had 

```
Uncaught SyntaxError: import declarations may only appear at top level of a module
```

But without that, I am not sure how to include the scss

**Comment by shankari** [comment]
I notice that none of the sample JSX for use with script tags, either in ngReact or in the basic react code
https://gist.githubusercontent.com/gaearon/c8e112dc74ac44aac4f673f2c39d19d1/raw/09b951c86c1bf1116af741fa4664511f2f179f0a/like_button.js
have any import code.

I suspect I need to remove all the imports if we want to use the `<script` tag approach without webpack. I will then need to compile the scss and add the relevant css as a stylesheet directly

**Comment by shankari** [comment]
ok so if I remove all the imports and include the css as a stylesheet directly, everything loads without any obvious errors.
HOWEVER, the scss files included in @kafitz's samples don't seem to compile correctly - in particular, running

```
$ npx node-sass examples/itinerum_tutorial/colors_v2.scss
/* Colors */
$
```

However, that is not my problem, and I am getting to the end of the weekend, so I am moving on to importing his code into a more complex angular app and then into e-mission.

**Comment by shankari** [comment]
actually, I'm just going to import it into e-mission directly. Note that although this works, I have not yet checked it in to my ngReact fork and will not check it in to e-mission, since the original code was shared as part of a secret gist by @kafitz.
@kafitz we should discuss how you and @jruzekowicz can see the examples.

**Comment by shankari** [comment]
Finally! I got the import into e-mission to work!
@kafitz @jruzekowicz, please see attached animated gif
![Screen Recording 2021-07-11 at 10 22 23 PM mov](https://user-images.githubusercontent.com/2423263/125235175-e59a5280-e296-11eb-922a-88aa0d3f1977.gif)

I basically reused the `<tutorial>` tag that I created with ngReact by copying that code over.
I then added a new template `www/templates/intro/survey.html`

```
<ion-content class="has-footer">
    <tutorial/>
</ion-content>
<ion-footer-bar class="no-bgColor" id="intro-footer">
  <button class="button button-block button-outline button-positive" ng-click="next()">
    SUBMIT
  </button>
</ion-footer-bar>
```

and added it as a slide right after this line
https://github.com/e-mission/e-mission-phone/blame/master/www/templates/intro/intro.html#L8

```
      <ion-slide>
        <ng-include src="consentFile"></ng-include>
      </ion-slide>
      <ion-slide>
        <ng-include src="'templates/intro/survey.html'"></ng-include>
      </ion-slide>
      <ion-slide>
        <ng-include src="explainFile"></ng-include>
      </ion-slide>
```

So we can now see the survey. We should still experiment with showing it in an ionic popup or popover if we need it, but a full screen display seems to be fine. We can also create a bi-directional binding similar to the examples in ngReact.

The current template shared by @kafitz doesn't have any variables, so I wasn't able to test out the bi-directional bindings, but the ngReact examples are pretty clear that it works.

**Comment by jruzekowicz** [comment]
Looking through these comments now. @shankari 

What else needs to be done today/tomorrow on the server side of things before we move on? 
@kafitz I looked through the code on your end and it appears you have a way to pull a list of the URLs already? What more can I help with on that end?

**Comment by shankari** [comment]
@jruzekowicz also based on our progress so far, I don't think you will need to make any native code changes. So you can use the devapp aka https://github.com/e-mission/e-mission-phone/#updating-the-ui-only to set up the phone code.

**Comment by kafitz** [comment]
@jruzekowicz It's pretty close now, I was thinking it should be made consistent with the export_stage of the pipeline (env variable names, default data path). Then it should just be run alongside it to check that it's seeing the data archives between both containers and able to generate some URLs in a JSON response. 

If that all works, I was thinking we could have 2 tests that (using a temporary configuration/data):
- check that these URLs are generated as expected
- received files are cleaned up

There's a basic tests.py file in the server repo now that could be used as a starting off point. If switching the mobile is a better use of time now, I can wrap all this up on my end.

**Comment by jruzekowicz** [comment]
@shankari I was able to get the phone install, Updating the e-mission-* plugins or adding new plugins, over the weekend. Is this still a valid install? Or do I need to do the ui only? 
Additionally, I know you said something last week about incorporating the export pipeline into a timed one hour export interval similar to other pipelines. Am I in change or handling this, and where should I start if so?
@kafitz Did we handle the "queue-like structure"? I know we said before that the data folder will act like a queue where we can pull out the data files in the order they were Witten to the folder based on their names. Does this still play into our implementation? If so where and does it still need to be done?

**Comment by shankari** [comment]
@jruzekowicz, that is a valid install; it just might be a bit slow to rebuild the app every time you make a change. using the devapp allows instant reload of the UI to see your changes. I would spend a time-bounded (~ 1 hour) effort to set up the UI only **in a separate directory**. The amortized cost of quicker UI changes is likely to be worth it.

**Comment by shankari** [comment]
My suggestion on how to proceed is as follows:
- I will create a simple data structure for the survey result - e.g. JSON and boolean flag for `isValid`
- I will pass in that data structure to the tutorial similar to the example with ngReact
- @kafitz can use that to maybe have a two page survey component (very basic) that fills in the values so we can confirm the two-way binding within the actual app

So far, we have just been copying the JSX files directly into the e-mission repo and running babel directly to create the js

Then, we break that dependency by having a separate repo that has all of the survey code, uses whatever building technique it wants and generates a (potentially bundled webpack) js file for me to import. We ensure that the two page survey component works with the imported js file

Then, @kafitz can happily work on the component independently and periodically check that the integration works.

**Comment by jruzekowicz** [comment]
https://community.unbounce.com/t/disabled-greyed-out-form-submit-button/12882
https://techstacker.com/how-to-enable-and-disable-buttons-with-javascript/
http://www.angulartutorial.net/2017/10/disable-submit-button-until-all.html

I have been looking through these primitive examples and it appears we would try to do similar on a larger scale and with most input fields. 
A few questions for the implementation I would be working on: 

- Can you expand on a simple data structure for the survey result? In my head we would pull from the survey that exists outside the app, format it into a js file that displays the data on the UI interface, and "grey out" the submit button until all fields that are marked as "required" have been filled in. Is the data structure referred to then just the results of that survey? It seems like that would come later in the process. 
- This may be due to my lack of understanding of what data structure refers to in this exact context, but what is the desired format and data being passed in? Is this simply the survey questions, and the boolean to keep track of if they are filled? 
-

**Comment by jruzekowicz** [comment]
Additionally, I have successfully gotten the devapp up and running. I installed both ways, plugin modification and otherwise through two separate cloned repos. This way if I need to, I have both set up and ready to run. @shankari once you are able to push your code from the weekend, I will spend time looking through it, and hopefully getting a better feel for how to use these new languages (new to me). 
I have been going through some tutorials today but find I will probably learn more through code that interacts with the UI that I can now see!

**Comment by shankari** [comment]
@jruzekowicz wrt
https://github.com/e-mission/e-mission-docs/issues/643#issuecomment-878536543
please don't just search for tips on the internet; javascript changes quickly and in incompatible ways and not everything that you read will be correct. For example, the third link that you put there (the angular tutorial) uses a more recent version of angular, **not** use angular1, which means that it **won't work**

I would suggest looking at the examples below and from other code in the e-mission-phone repo. This is a fairly complex project and there are a ton of examples there already. That is likely to be much faster than digging through multiple incompatible "tutorials" on the internet.

**Comment by shankari** [comment]
Example in ngReact (very simple, please clone and experiment to be comfortable with 2-way binding):
https://github.com/shankari/ngReact/tree/master/examples/itinerum_tutorial

Integration with the e-mission codebase (more complicated, but still pretty straightforward if you understand the commit comments and the notes above):
https://github.com/shankari/e-mission-phone/tree/explore_react_integration

Note the updates to the README with additional instructions.
Also note the additional package dependencies - remember to either re-run setup, or manually run the setup steps (e.g. `npm install`) to ensure that the dependencies are updated.

Good luck!

**Comment by jruzekowicz** [comment]
Sounds good. Some questions I have on the commit listed here: https://github.com/shankari/e-mission-phone/commit/45b6461c5caaad6d2796e45613d4973a83c9177b#commitcomment-53429232

Additionally, we can talk about this during our meeting however I am still trying to understand the main task. 


> * Can you expand on a simple data structure for the survey result? In my head we would pull from the survey that exists outside the app, format it into a js file that displays the data on the UI interface, and "grey out" the submit button until all fields that are marked as "required" have been filled in. Is the data structure referred to then just the results of that survey? It seems like that would come later in the process.
> * This may be due to my lack of understanding of what data structure refers to in this exact context, but what is the desired format and data being passed in? Is this simply the survey questions, and the boolean to keep track of if they are filled?

Above were my questions from before that still stand. Not sure what creating a data structure for this context entails ect.

**Comment by shankari** [comment]
At this point we are only creating a placeholder data structure, @kafitz will fill it in later.
So something like:

```
$scope.surveyState = {
    valid: False,
    result: {}
}
```

will probably be good enough. This is pseudo-code, you may need to adjust or fix syntax.

**Comment by shankari** [comment]
The two players in the two-way binding are react and angular. The angular code (in `app.js` or `intro.js`) passes a `$scope.XXX` to the react component (e.g. `Hello.jsx` in the ngReact standard examples, `Tutorial.jsx` in the itinerum example`). The react component changes those values internally, but because of the two-way binding, the `$scope` variables in angular also see the changes.

I don't know how it works under the hood, but the overall effect is that that of shared variables between the angular code and the react code.

**Comment by jruzekowicz** [comment]
```
var app = angular.module('app', ['react']);

app.controller('mainCtrl', function ($scope) {
  $scope.person = { fname: 'Clark', lname: 'Kent' };
});

app.directive('tutorial', function (reactDirective) {
  return reactDirective(TutorialComponent);
});
```
Above is the current code within @shankari commit under the ngReact repo. I think this is a good setup for what I am trying to do with the data structure? 

Within the current code that @shankari added to the e-mission-phone repo, her newest addition included: 

```
.directive('tutorial', function (reactDirective) {
  return reactDirective(TutorialComponent);
})
```

To add in the data structure and boolean, I plan to make the following addition to that same file (www/js/app.js). 

```
.directive('tutorial', function (reactDirective) {
  return reactDirective(TutorialComponent);
})

.controller('mainCtrl', function ($scope) {
  $scope.surveyState = { valid: False; result : {} };
})
```
Pulling from my research: "A controller is a JavaScript Object, created by a standard JavaScript object constructor." Thus my goal here is to create an object that incorporates the survey state. This is similar and runs parallel to what happens with the name variable within the examples. 

How does this sound?

**Comment by shankari** [comment]
@jruzekowicz I don't think I can tell you exactly what to do here - that would be equivalent to just doing it. I would suggest looking at the history of my exploration with ngReact (including https://github.com/shankari/ngReact/tree/master/examples/jsx-transformer-precompiled) and fully understanding how they work, including by modifying them to experiment with what does and does not work.

The above code is not complete and will not pass in the values to the tutorial component. you can verify this by trying to print out the values in the tutorial component (like they do in hello).

**Comment by jruzekowicz** [comment]
Got it. That was my next question. I am a little confused about what "passing the values to the tutorial component" would mean? 

Looking at the files I am still trying to figure out where and how the name variable itself is passed around.

**Comment by jruzekowicz** [comment]
> @jruzekowicz I don't think I can tell you exactly what to do here - that would be equivalent to just doing it. I would suggest looking at the history of my exploration with ngReact (including https://github.com/shankari/ngReact/tree/master/examples/jsx-transformer-precompiled) and fully understanding how they work, including by modifying them to experiment with what does and does not work. e.g. if you pass in the structure above to `Hello.jsx`, does it work?
> 
> The above code is not complete and will not pass in the values to the tutorial component correctly. Again, you can verify this by trying to print out the values in the tutorial component (like they do in hello).

Good Morning, @shankari I am attempting to modify these files and add in structures I have developed (first in the ngReact setting, then plan to shift over to the e-mission-phone. I have made edits but they do not seem to be reflected when I run the server. Am I missing a step to sync everything together? Do I need to push a commit? Not familiar with how the server is being hosted and from where exactly it pulls the information (I though from my project folder directly).

**Comment by jruzekowicz** [comment]
For future use and testing: 
Note that in browsers such as Chrome and Safari, a cached memory of the served up page will likely be stored. To see immediate changes reflected, empty the cache and reload the page.

**Comment by jruzekowicz** [comment]
@shankari A first commit in my attempts. Still trying to figure out what is missing here. 
https://github.com/jruzekowicz/e-mission-phone/commit/2690c40b5ece8c9922b76429e3392b89988bfc4f 

Additionally, would you like this opened as a pull request and if so to what branch?

**Comment by shankari** [comment]
So one high level question is: where do we retrieve the schema for this survey from?
- angular code retrieves it, and passes it in to the react component
- react component retrieves it, potentially with the angular component passing in the URL of the survey

@kafitz how do you do this now? How does the app know which survey to pull?

**Comment by shankari** [comment]
Theoretically, the react component can do everything that the existing native app does wrt survey. It can popup a dialog asking for a code, and then use that code to retrieve the survey JSON from the server and display it.

However, we will start with the angular integration because then we can continue to use the QR code-based approach to selecting a survey, which has the benefit of not entering the code manually, and also can pass in other arguments to the app/survey potentially.

**Comment by shankari** [comment]
> Additionally, would you like this opened as a pull request and if so to what branch?

Yes, PR would be good because then I can put suggestions directly into the review.
And it would be to the new branch that I created (https://github.com/shankari/e-mission-phone/tree/explore_react_integration) in my fork

**Comment by shankari** [comment]
would we want to display/launch surveys from the trip diary? and would those surveys be the same component with a different JSON, or a separate component?

**Comment by shankari** [comment]
The UNSW team (@asiripanich) has a version that displays surveys created using Enketo Toolbox. There is also Survey.js which has a free survey builder and a bunch of libraries to display the survey, including in react. They are excitability designed for integration into other websites, so may be easier to use than Enketo.

We may want to work together to define survey use in the platform, standardize on the workflow (e.g. retrieve survey JSON, pass it in to component, display component, retrieve full and summary results), and then we can have components for a few survey integrations.

**Comment by kafitz** [comment]
Currently I'm experiencing trouble with the final step of integrating a Typescript React component into ngReact. The first thing I've found was babel v6 does not seem to support TypeScript, only 7 and above. I've upgraded it for my test, but is babel v6 used elsewhere within the e-mission-phone build stage?

I've forked my changes here: https://github.com/kafitz/e-mission-phone/tree/explore_react_integration/www/components/itinerum_questionnaire as well as importing this code in index.html and adding a directive (same as your Tutorial) to app.js. When I compare my generated .js files between my new `itinerum_questionnaire` component and your working `itinerum_tutorial`, both seem to contain React-specific functions from transpiling, but it seems like Angular isn't able to import mine.

I'm continuing to hash it out, but was wondering if @shankari had any suggestions about how to proceed since it felt like I was sputtering a bit.

**Comment by shankari** [comment]
> is babel v6 used elsewhere within the e-mission-phone build stage?

e-mission build stage does not do anything sophisticated at this point since we don't use typescript :)
We assume that all files are javascript and in `www/js` or installed into `www/lib` using bower.

> but it seems like Angular isn't able to import mine.

I don't have time to run this right now, so could you elaborate on your error a bit more so I can respond quickly if sometime jumps out at me.

**Comment by kafitz** [comment]
Ultimately I'm trying to just display "Hello World" right with now with a very simple component: https://gist.github.com/kafitz/ee81d6a76d5f76ba432b01c23f6738e1 (gist put them in alphabetically order, but I meant to show my QuestionStack.tsx -> QuestionStack.js -> Tutorial,js

Judging by the output, it seems to Typescript stage of babel is working fine, but transforming it into ngReact doesn't seem to yield all the additional boilerplate and helper functions that yours does. I never totally grasped the difference between modules.export/AMD importing syntax in JS so I didn't know if it could be that.

In addition to adding the directive with the name string "questionnaire", I also updated the `app/templates/intro/survey.html` to show `<questionnaire />`.

In the app.js directive, I tried logging the component object itself and receive the screenshotted error when I reach the questionnaire (formerly tutorial) screen within the app.

![directive](https://user-images.githubusercontent.com/253651/128055180-1ba3ee50-360b-47e2-bee1-5b3ff7a64fa2.png)

**Comment by kafitz** [comment]
I'm starting to think this more has to do with my React code itself since I've started using their functional components over extending the React.Component class. Although ngReact returns React.Element() from my .tsx, it might not be sufficient for passing to Angular in this case.

Will keep researching to see if I should refactor the code or if there's some additional configuration for babel.

**Comment by shankari** [comment]
@kafitz

- can you use the non-minified version of the react libraries so we can get a better sense of what is going on?

```
diff --git a/www/index.html b/www/index.html
index 2dc363b7..2c9e7414 100644
--- a/www/index.html
+++ b/www/index.html
@@ -61,9 +61,9 @@
     <!-- Manually installed javascript for libraries that are not published through bower -->
     <!-- cordova script (this will be a 404 in the browser) -->
     <script type="text/javascript" src="cordova.js"></script>
-    <script src="lib/react/react.min.js"></script>
-    <script src="lib/react/react-dom.min.js"></script>
-    <script src="lib/ngReact/ngReact.min.js"></script>
+    <script src="lib/react/react.js"></script>
+    <script src="lib/react/react-dom.js"></script>
+    <script src="lib/ngReact/ngReact.js"></script>

     <!-- React components -->
     <script src="components/itinerum_tutorial/Tutorial.js"></script>
```
 then send me the new stacktrace and I will take a look at the `ngReact` code to see if I can get a clue.

**Comment by kafitz** [comment]
I've swapped out to the non-minified version, but unfortunately get the same stack track as my screenshot above.
```bash
[phonegap] [console.error] Error: ReactComponent name attribute must be specified
[phonegap]     at getReactComponent (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ngReact/ngReact.js:35:13)
[phonegap]     at link (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ngReact/ngReact.js:262:32)
[phonegap]     at invokeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:22993:9)
[phonegap]     at nodeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:22392:11)
[phonegap]     at compositeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:21703:13)
[phonegap]     at compositeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:21706:13)
[phonegap]     at nodeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:22387:24)
[phonegap]     at compositeLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:21703:13)
[phonegap]     at publicLinkFn (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:21583:30)
[phonegap]     at link (http://localhost/_app_file_/data/user/0/edu.berkeley.eecs.emission.devapp/files/phonegapdevapp/www/lib/ionic/js/ionic.bundle.js:39027:38) <questionnaire>
```

To get a sense of what was going on, I re-pulled your original exploration branch and confirmed that to work with my environment. Then I commented out my edits to match yours with a diff tool and made sure my own branch was working again with your example.

This leads me to two places: the app,js directive (perhaps I should be doing this a different way; originally I thought my issue was chaining accidentally chaining them but that hasn't fixed my error). Since the `ReactComponent name attribute must be specified` error shows up pretty early in the ngReact code (https://ngreact.github.io/ngReact/docs/ngReact.html), the other possibility is my functional component (I haven't found much by Googling "ngreact functional component") not provide this "name" attribute to Angular somehow.

It would much more copy & paste to be able to use my functional components, but I'll rewrite as classes if that's advised at this point.

**Comment by shankari** [comment]
Looking through the ngReact code, it looks like the error is from 

```
 33     // a React component name must be specified
 34     if (!name) {
 35       throw new Error('ReactComponent name attribute must be specified');
 36     }
 37
```

which is in turn called from 

```
256   var reactDirective = function($injector) {
257     return function(reactComponentName, staticProps, conf, injectableProps) {
258       var directive = {
259         restrict: 'E',
260         replace: true,
261         link: function(scope, elem, attrs) {
262           var reactComponent = getReactComponent(reactComponentName, $injector);
```

The part that is confusing me is that at that point, ngReact hasn't even started looking up the component. It is just seeing if the name exists or not. @kafitz can you add breakpoints or log statements to the ngReact code to figure out what is going on? 

Maybe something like

```
--- /Users/kshankar/OSS/ngReact/ngReact.js	2021-07-11 10:39:22.000000000 -0700
+++ www/lib/ngReact/ngReact.js	2021-08-04 10:55:48.000000000 -0700
@@ -26,10 +26,13 @@
   // available on window
   function getReactComponent( name, $injector ) {
     // if name is a function assume it is component and return it
+    console.log("Trying to get react component with name '"+name+"'");
     if (angular.isFunction(name)) {
+      console.log("It is a function, returning the name directly '"+name+"'");
       return name;
     }

+    console.log("It is not a function, checking whether it exists '"+name+"'");
     // a React component name must be specified
     if (!name) {
       throw new Error('ReactComponent name attribute must be specified');
```

**Comment by shankari** [comment]
FYI, I ran with this logging enabled for only the tutorial component, and I got the following logs

```
[phonegap] [console.log] Trying to get react component with name 'function TutorialComponent() {
[phonegap]         _classCallCheck(this, TutorialComponent);
[phonegap]
[phonegap]         return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).apply(this, arguments));
[phonegap]     }'
[phonegap] [console.log] It is a function, returning the name directly 'function TutorialComponent() {
[phonegap]         _classCallCheck(this, TutorialComponent);
[phonegap]
[phonegap]         return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).apply(this, arguments));
[phonegap]     }'
```

So that's why it doesn't fail for the TutorialComponent; it is because it is already a function.

Looking at the autogenerated `Tutorial.js`, I see the same function defined 

```
    function TutorialComponent() {
        _classCallCheck(this, TutorialComponent);

        return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).apply(this, arguments));
    }
```

Will have to see if @kafitz's new component generates something similar.

**Comment by shankari** [comment]
@kafitz I tried to compile your script using the babel command line at:
https://github.com/kafitz/e-mission-phone/commit/4e2f853a59372aa5bf2c71c92133cbf4ac3c73fd#diff-d5e0313ae90d6090ceb6d48af20e5cc300bc59403bcfc00c7f788fa0632c318dR1

I think I have all the correct versions installed (I had to guess, but got the most recent versions), but I still get the following error.

```
NODE_ENV=production npx babel www/components/itinerum_questionnaire/ --out-dir www/components/itinerum_questionnaire/ --extensions ".ts,.tsx" --presets @babel/preset-env,@babel/preset-typescript,react-app --plugins @babel/plugin-proposal-class-properties
Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/kshankar/e-mission/upgrade_upload_process/node_modules/babel-preset-react-app/index.js
    at createDescriptor (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/@babel/core/lib/config/config-descriptors.js:211:11)
    at createDescriptor.next (<anonymous>)
    at step (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/gensync/index.js:261:32)
    at evaluateAsync (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/gensync/index.js:291:5)
    at /Users/kshankar/e-mission/upgrade_upload_process/node_modules/gensync/index.js:44:11
    at Array.forEach (<anonymous>)
    at Function.async (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/gensync/index.js:43:15)
    at Function.all (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/gensync/index.js:216:13)
    at Generator.next (<anonymous>)
    at createDescriptors (/Users/kshankar/e-mission/upgrade_upload_process/node_modules/@babel/core/lib/config/config-descriptors.js:142:41)
```

**Comment by shankari** [comment]
Ah, based on 
https://stackoverflow.com/a/58676558/4040267

changed the command line to 

NODE_ENV=production npx babel www/components/itinerum_questionnaire/ --out-dir www/components/itinerum_questionnaire/ --extensions ".ts,.tsx" --presets @babel/preset-env,@babel/preset-typescript --plugins @babel/plugin-proposal-class-properties

(removed `react-app`) and it works now.

**Comment by shankari** [comment]
@kafitz are you sure you are able to run with the `react_app` babel plugin?

Note that for my transpiling, I used

```
npx babel www/components/itinerum_tutorial/ --out-dir www/components/itinerum_tutorial/ --presets react-app/prod
```

This used to work just fine with babel 6.x, but it fails with babel 7.x. 

```
$ npx babel --version
7.14.8 (@babel/core 7.14.8)
$ npx babel www/components/itinerum_tutorial/ --out-dir www/components/itinerum_tutorial/ --presets react-app/prod
Error: Cannot find module 'react-app/prod'
```

Going back to 6.x, it works again

```
$ npx babel --version
6.26.0 (babel-core 6.26.3)
$ npx babel www/components/itinerum_tutorial/ --out-dir www/components/itinerum_tutorial/ --presets react-app/prod
www/components/itinerum_tutorial/Tutorial.js -> www/components/itinerum_tutorial/Tutorial.js
www/components/itinerum_tutorial/Tutorial.jsx -> www/components/itinerum_tutorial/Tutorial.js
```

You saw that I had to remove the `react-app` plugin to get your command line to work
https://github.com/e-mission/e-mission-docs/issues/643#issuecomment-892924322

How did you manage to get that preset to work? I want to check the output with that included.

**Comment by kafitz** [comment]
I was digging into the differences of the presets earlier, it looks like the babel-preset-react-app that you had includes some that are now part of the normal @babel/preset-react, so I had been experimenting with that in --plugins (I think we still need one of them). Both seem to yield similar outputs.

My current package,json looks like this:
```
  "devDependencies": {
    "@ionic/cli": "6.10.1",
    "bower": "1.8.8",
    "cordova": "9.0.0",
    "phonegap": "9.0.0+cordova.9.0.0"
  },
  "dependencies": {
    "@babel/cli": "^7.14.8",
    "@babel/core": "^7.14.8",
    "@babel/preset-env": "^7.14.9",
    "@babel/preset-react": "^7.14.5",
    "@babel/preset-typescript": "^7.14.5",
    "fs-extra": "^9.0.1",
    "klaw-sync": "^6.0.0"
  }
```

The significant difference I see in the output is that function you called attention to:
```
var TutorialComponent = function (_React$Component) {
    _inherits(TutorialComponent, _React$Component);

    function TutorialComponent() {
        _classCallCheck(this, TutorialComponent);

        return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).apply(this, arguments));
    }

    _createClass(TutorialComponent, [{
```

My version looks like this after running babel:
```
var ItinerumQuestionStack = function ItinerumQuestionStack(props) {
  return /*#__PURE__*/React.createElement("div", null, "HELLO WORLD!");
};
```
and seems to be missing most of the guts from the class version. It's odd because I did found some issues threads that indicate this _should_ be possible:
- https://github.com/ngReact/ngReact/issues/142
- https://github.com/ngReact/ngReact/issues/172

**Comment by shankari** [comment]
FYI: I got the `react-app/prod` command directly from the react website, so I don't think it is completely obsolete.
https://reactjs.org/docs/add-react-to-a-website.html#run-jsx-preprocessor

If you remove the parts of your code that require babel 7 and run the exact same command line that I use, do you get the same class structure? if so, then we know that the problem is with babel 6 v/s babel 7 and can figure out which way to go forward.

**Comment by kafitz** [comment]
The issue with v6 is I didn't think Typescript had a compatible preset so I get stuck there first.

**Comment by shankari** [comment]
Looks like this works with CRA
(https://github.com/facebook/create-react-app/issues/5443)
so maybe see how they are using babel under the hood?

**Comment by kafitz** [comment]
Progress! Although I'm not entirely certain how to get Babel to fully cooperate yet...if I remove the various import/exports from my generated QuestionLayout.js, I see my "HELLO WORLD!" text.

```
// "use strict";

// Object.defineProperty(exports, "__esModule", {
//   value: true
// });
// exports.ItinerumQuestionStack = void 0;

/* 
 * www/itinerum_questionnaire/QuestionStack.tsx
 */
var questionMap = {
  dropdown: undefined,
  checkbox: undefined,
  number: undefined,
  address: undefined,
  textBox: undefined
};

var ItinerumQuestionStack = function ItinerumQuestionStack(props) {
  return /*#__PURE__*/React.createElement("div", null, "HELLO WORLD!");
};

// exports.ItinerumQuestionStack = ItinerumQuestionStack;
```

This was generated by the babel command:
```
NODE_ENV=production npx babel www/components/itinerum_questionnaire/ --out-dir www/components/itinerum_questionnaire/ --extensions ".ts,.tsx"```
and a `.babelrc` file that I was temporarily using:
```
{
  "presets": [
    "@babel/preset-env",
    "@babel/preset-typescript",
    [ "react-app", { "absoluteRuntime": false } ]
  ],
  "plugins": [
    "@babel/plugin-proposal-class-properties" 
  ]
}
```

**Comment by kafitz** [comment]
Unfortunately there's still something causing it to break whenever I attempt to include an import from node_modules.

Such as:
```
/* 
 * www/itinerum_questionnaire/QuestionStack.tsx
 */
import React, { useRef } from 'react';

import { StackQuestion } from './questionnaireTypes';


const questionMap: {[key: string]: any} = {
    dropdown: undefined,
    checkbox: undefined,
    number: undefined,
    address: undefined,
    textBox: undefined,
};


export interface QuestionStackProps {
    questions: StackQuestion[];
}

const ItinerumQuestionStack = (props: QuestionStackProps) => {
    const stackRef = useRef(null);

    return (
        <div>
            abc abc
        </div>
    );
}
```

generating:
```
// "use strict";

var _typeof = require("@babel/runtime/helpers/typeof");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* 
 * www/itinerum_questionnaire/QuestionStack.tsx
 */
var questionMap = {
  dropdown: undefined,
  checkbox: undefined,
  number: undefined,
  address: undefined,
  textBox: undefined
};

var ItinerumQuestionStack = function ItinerumQuestionStack(props) {
  var stackRef = (0, _react.useRef)(null);
  return /*#__PURE__*/_react["default"].createElement("div", null, "abc abc");
};
```

Continuing to investigate.

**Comment by shankari** [comment]
@kafitz did you see the example with es6 and webpack (https://github.com/ngReact/ngReact/tree/master/examples/es6-tests-webpack). That does seem to have some imports e.g. https://github.com/shankari/ngReact/blob/master/examples/es6-tests-webpack/lib/watch-list-component.js#L1

I didn't try to run that, since it didn't seem to be needed. LMK if that is closer to what you need, and I can see whether it works.

**Comment by shankari** [comment]
@kafitz not sure if this helps, but I have some additional information. I basically added all the tutorial stuff back as well so I could compare the tutorial and the react components side by side.

I first realized that the `reactComponentName` passed in was the full function for the tutorial component and undefined for the questionnaire

```
Creating directive with name = 'function TutorialComponent() {
        _classCallCheck(this, TutorialComponent);

        return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).apply(this, arguments));
    }'


Creating directive with name = 'undefined'
```

I then tried accessing the variables we pass in to the directives. TutorialComponent is a function

```
TutorialComponent
ƒ TutorialComponent() {
        _classCallCheck(this, TutorialComponent);

        return _possibleConstructorReturn(this, (TutorialComponent.__proto__ || Object.getPrototypeOf(TutorialComponent)).appl…
```

but the react component is `undefined`

```
ItinerumQuestionStack
undefined
```

do you have to do anything special to "export" the `ItinerumQuestionStack`?

**Comment by kafitz** [comment]
That example from two comments back seems to be what I'm looking for, but haven't managed it myself yet. Babel is fine when I import from a sibling file (since it generates that files and removes the import), so I think it's generating an import/require() for the 3rd-party libraries that ngReact isn't liking.

I also noticed similar regarding the reactComponentName being the full function definition, so it made me slightly more confused how much stateless component worked (will still have to log that and see).

Generally speaking, I would a `export default ItinerumQuestionStack;` at the bottom of my component files which is what I thought could be causing these problems initially. I notice they have it still in your example above. When I manually commented out the `use strict` at the top of my generated files and that bottom export, that's when it did seem to work.

**Comment by kafitz** [comment]
Primarily my reservations have been having to completely re-implement my state handling and JSON logic. My original JS-only version was too messy to consider here, so this TypeScript version was a complete rewrite. So long as we can continue to use TypeScript, then it probably won't be such a big deal to reformat the state components as class style ones.

**Comment by shankari** [comment]
There are a couple of other things we can try. The issue that you linked to a couple of comments above said something like 

> The interesting thing is that <react-component syntax works just fine.

So we could try the directive and see what it gives us.
https://github.com/ngReact/ngReact#the-react-component-directive

I would also suggest first trying to get stateless components to work with the ngReact examples, since there are code examples there and it is faster to develop directly in the browser.

We can also pass the name of the component 
> Alternatively you can provide the name of the component

which will trigger the alternate code path where ngReact tries to use `$injector.get` with it.

**Comment by kafitz** [comment]
I've still not gotten to the root of the issue but have committed my latest half-working state here: https://github.com/kafitz/e-mission-phone/ under the same `explore_react_integration` branch. 

I've just noticed that the package.json is not committed by default since it must be generated off package.serve.json, so I'll include mine here for now:
```
{
  "name": "edu.berkeley.eecs.emission",
  "version": "2.5.0",
  "displayName": "emission",
  "license": "BSD-3-Clause",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/e-mission/e-mission-phone.git"
  },
  "scripts": {
    "setup-serve": "./bin/download_settings_controls.js && ./bin/setup_autodeploy.js",
    "compile-react": "babel www/components/itinerum_tutorial/ --out-dir www/components/itinerum_tutorial/ --presets react-app/prod",
    "serve": "phonegap --verbose serve"
  },
  "devDependencies": {
    "@ionic/cli": "6.10.1",
    "bower": "1.8.8",
    "cordova": "9.0.0",
    "phonegap": "9.0.0+cordova.9.0.0"
  },
  "dependencies": {
    "@babel/cli": "^7.14.8",
    "@babel/core": "^7.15.0",
    "@babel/preset-env": "^7.14.9",
    "@babel/preset-typescript": "^7.14.5",
    "@babel/runtime": "^7.14.8",
    "@material-ui/core": "^4.12.3",
    "babel-preset-react-app": "^10.0.0",
    "fs-extra": "^9.0.1",
    "klaw-sync": "^6.0.0"
  }
}
```

The only files I've been editing are the `app.js`, the `survey,html` and the .tsx files within that `www/components/itinerum_questionnaire` directory. That directory contains a very basic `ItinerumQuestionStack.tsx` that I've been trying to get to work first with its imports, and a `ItinerumQuestionStack.orig.tsx` which was originally supposed to be my reference example. The only additional file will be the tsconfig.json in the repo root.

For compiling, my command is:
```
NODE_ENV=production npx babel www/components/itinerum_questionnaire/ --out-dir www/components/itinerum_questionnaire/ --extensions ".ts,.tsx" --presets @babel/preset-env,@babel/preset-typescript,react-app --plugins @babel/plugin-proposal-class-properties
```
which is also in a `compile.txt` file within my `itinerum_questionnaire` dir.

I think part of my slowdown is not knowing how to debug off the phone so I'm often making changes, reloading em-dev and seeing if it worked. Is there a better way I should be doing it?

**Comment by shankari** [comment]
@kafitz that's why I think it might be easier to work off the ngReact repo first, since there is no phone component involved. At this point, I think that if it works in ngReact, we can get it to work on the phone.

**Comment by kafitz** [comment]
I see, I didn't think of trying to use it independently so far. So that would require having a demo Angular app that works in the browser (for now) and testing my .tsx component there first? I'll start doing that next

**Comment by shankari** [comment]
the examples in ngReact already have a demo angular app. I would start by copying https://github.com/shankari/ngReact/tree/master/examples/itinerum_tutorial and going from there.

**Comment by kafitz** [comment]
Current status / what I've tried:
- All combinations of babel7 and react plugins that I can find
- use ngReact example to debug: all errors pertain to "exports" or "require()" not found
- trying as many possible options for tsconfig.json as could be located online
- online indicated that babelify/gulp/webpack needed to brings these newer versions of package imports to browser
- looking at ngReact webpack examples did not show me much on how babel is actually called, decided to include webpack
- bower seems to only supply React up to 0.16.1; need at least 0.16.8 (written with/possibly 17)
- use create-react-app to generate a working typescript/react build chain; thinking is to compile JS with that and copy as needed into e-mission-phone `www` folder
    - added react-app-rewired to easily customize webpack.config.json
    - this at least yielded me a working of example of my stateless React component with imports
    - now able to run ngReact + ItinerumQuestionStack without console errors, but cannot see `ItinerumQuestionStack` object in app.js (for ngReact directive)
        - assumption is this has to do with how the React component is initially loaded by create-react-app's index.tsx. This is fine when loading the React component when mounting an index.html element like normal, but stuck on getting app.js to see my component's object variable.

Basic `create-react-app` example: https://github.com/kafitz/itinerum-questionnaire

**Comment by shankari** [comment]
@kafitz that's great progress. @jruzekowicz is making good progress on the last part of the integration (saving to DB and pushing up to server), so might be able to help with this soon as well in case having another person trying things out in parallel would help.

**Comment by shankari** [comment]
@kafitz have you made any progress on this? I might have some time this weekend to work on this, although my javascriot-fu is pretty week

**Comment by shankari** [comment]
or are you giving up on the idea of stateless components and planning to use stateful componets

**Comment by kafitz** [comment]
I have the stateless components working here: https://github.com/hexmap/emission-plugin-itinerum-questionnaire I worked off the ngReact es6 example using Webpack/babel, but everything at the latest versions.

Most of my trouble this week was in passing my JSON schema from Angular->React, but I think I'd been unnecessarily caught up trying to emulate the Promises to keep it true to the ngReact example. This is finished, so now I'm working on merging my React dashboard code to parse and manage the survey with my Flutter version of the Itinerum mobile survey.

So far no big issues on that, the next big test will be when I implement navigation between survey questions.

**Comment by kafitz** [comment]
The thing that finally got the stateless version to work was having this factory function: https://github.com/hexmap/emission-plugin-itinerum-questionnaire/blob/master/lib/itinerum-survey-component.tsx#L23

Why this works, however, is still a mystery.

**Comment by shankari** [comment]
Itinerum is currently on hiatus. We have integrated an enketo toolbox-based survey for the initial survey 
https://github.com/e-mission/e-mission-docs/issues/727
